### PR TITLE
KAFKA-2962: stream-table table-table joins

### PR DESCRIPTION
**Pull request by ymatsuda**
@guozhangwang

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 1]
```diff
@@ -172,4 +172,14 @@
      */
     void process(ProcessorSupplier<K, V> processorSupplier, String... stateStoreNames);
 
+    /**
+     * Combines values of this stream with KTable using Left Join.
+     *
+     * @param ktable the instance ogf KTable joined with this stream
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: ogf => of.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.streams.kstream.KStreamBuilder;
+import org.apache.kafka.streams.kstream.ValueJoiner;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class AbstractStream<K> {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:27]: Why do we want to define it as abstract class?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 3]
```diff
@@ -244,4 +238,17 @@ public void process(final ProcessorSupplier<K, V> processorSupplier, String... s
...
     }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <V1, R> KStream<K, R> leftJoin(KTable<K, V1> other, ValueJoiner<V, V1, R> joiner) {
+        Set<String> allSourceNodes = ensureJoinableWith((AbstractStream<K>) other);
+
+        String name = topology.newName(LEFTJOIN_NAME);
+
+        topology.addProcessor(name, new KStreamKTableLeftJoin<>(((KTableImpl<K, ?, V1>) other).valueGetterSupplier(), joiner));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: Do we miss the parent name here on purpose?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java (hunk 2 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractStream.java:27]: This is the place for common code of KStreamImpl and KTableImpl. It should not be instantiated by itself.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 4]
```diff
@@ -44,13 +45,22 @@
...
+
+    public static final String JOINOTHER_NAME = "KTABLE-JOINOTHER-";
+
+    public static final String OUTERTHIS_NAME = "KTABLE-OUTERTHIS-";
+
+    public static final String OUTEROTHER_NAME = "KTABLE-OUTEROTHER-";
+
+    public static final String LEFTJOIN_NAME = "KTABLE-LEFTJOIN-";
+
+    public static final String MERGE_NAME = "KSTREAM-MERGE-";
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: `KTABLE-MERGE-`?

**Comment by guozhangwang** [comment]
LGTM. Jenkins failures seem not relevant.

**Pull request closed without merging**